enterprise/utils.py

```diff
@@ -1108,3 +1110,103 @@ def get_platform_logo_url():
+                user_item = TSC.UserItem(username, TSC.UserItem.Roles.Viewer)
+                user = server.users.add(user_item)
+                enterprise_analytics_user = EnterpriseAnalyticsUser.objects.create(
+                    enterprise_customer_user=enterprise_customer_user,
+                    analytics_user_id=user.id
+                )
+                enterprise_analytics_user.save()
+                add_user_to_tableau_group(user.id, str(enterprise_uuid))
+        except ServerResponseError as exc:
+            LOGGER.error(
+                '[TABLEAU USER SYNC] Could not sync enterprise admin user %s in tableau.'
+                'Server returned: %s', username, str(exc),
+            )
+
+
+def get_tableau_server():
+    """
+    Return Tableau ServerAuth and Server instances or None
+    """
+    try:
+        tableau_auth = TSC.TableauAuth(settings.TABLEAU_ADMIN_USER, settings.TABLEAU_ADMIN_USER_PASSWORD)
+        server = TSC.Server(settings.TABLEAU_URL)
+        return tableau_auth, server
+    except AttributeError as err:
+        LOGGER.error(
+            '[TABLEAU USER SYNC] Could not sync enterprise admin user in tableau.'
+            'Error detail: %s', str(err),
+        )
+        return None, None
+
+
+def add_user_to_tableau_group(tableau_user_id, tableau_group):
+    """
+    Associate a user with the group in tableau. Create the group if it does not already exist.
+    """
+    tableau_auth, server = get_tableau_server()
+    if tableau_auth and server:
+        try:
+            with server.auth.sign_in(tableau_auth):
+                group_to_create = TSC.GroupItem(tableau_group)
+                group = server.groups.create(group_to_create)
+                server.groups.add_user(group, tableau_user_id)
+        except ServerResponseError as exc:
+            if exc.code == '409009':  # Group already exists
+                try:
+                    with server.auth.sign_in(tableau_auth):
+                        enterprise_group = TSC.GroupItem(tableau_group)
+                        server.groups.add_user(enterprise_group, tableau_user_id)
+                except ServerResponseError as exc:
+                    LOGGER.error(
+                        '[TABLEAU USER SYNC] Could not associate user %s with group %s in tableau.'
+                        'Server returned: %s', tableau_user_id, tableau_group, str(exc),
+                    )
+            else:
+                raise
+
+
+def delete_tableau_user(enterprise_customer_user):
+    """
+    Delete the user in tableau.
+    """
+    from enterprise.models import EnterpriseAnalyticsUser  # pylint: disable=import-outside-toplevel
+    enterprise_analytics_user = EnterpriseAnalyticsUser.objects.filter(
+        enterprise_customer_user=enterprise_customer_user
+    ).first()
+    if enterprise_analytics_user:
+        delete_tableau_user_by_id(enterprise_analytics_user.analytics_user_id)
+    else:
+        LOGGER.error(
```

Comment:
Should this be `warn` rather than `error`?

Reply:
Done